add mass operation for assign, unassign, take-leave, cancel-leave

sigmund-c opened the pull request:
resolves #78 
Usage:
`assign s/1 w/1 Cashier w/2 Pilot w/3 Janitor`
`unassign s/2 w/1 w/2 w/3`

sigmund-c committed:
add mass op for assign, unassign, take-leave, cancel-leave

sigmund-c committed:
add mass op for assign, unassign, take-leave, cancel-leave

sigmund-c committed:
add DG for mass ops

sigmund-c committed:
add newline at EoF

wakululuu commented on docs/DeveloperGuide.md:
```diff
@@ -304,6 +304,39 @@ from the 1st shift in the McScheduler. The `unassign` command creates a dummy `A
+
+### MassOps Feature
+For certain commands that will be frequently used (`assign`, `unassign`, `take-leave`, `cancel-leave`), mass
+ operations are supported to reduce the required number of command calls.
```
```suggestion
operations are supported to reduce the required number of command calls.
```

wakululuu commented on docs/DeveloperGuide.md:
```diff
@@ -304,6 +304,39 @@ from the 1st shift in the McScheduler. The `unassign` command creates a dummy `A
+ 
+#### Implementation
+These operations consist of their own Command class and CommandParser class. In each of the supported 
+CommandParser class, Mass operations uses `ParserUtil`'s `GetAllValues()` method, which parses the user input 
```
```suggestion
`*CommandParser` class, mass operations uses the `ArgumentMultimap#getAllValues(Prefix)` method, which parses the user input 
```

wakululuu commented on docs/DeveloperGuide.md:
```diff
@@ -304,6 +304,39 @@ from the 1st shift in the McScheduler. The `unassign` command creates a dummy `A
+ operations are supported to reduce the required number of command calls.
+ 
+#### Implementation
+These operations consist of their own Command class and CommandParser class. In each of the supported 
```
```suggestion
These operations consist of their own `*Command` class and `*CommandParser` class. In each of the supported 
```

wakululuu commented on src/main/java/seedu/address/logic/commands/AssignCommand.java:
```diff
@@ -27,37 +26,33 @@
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds shift, worker(s) and role(s) assignment(s) "
+            + "to the McScheduler by the index numbers used in the last worker and shift listings. "
             + "\nParameters: "
-            + PREFIX_SHIFT + "SHIFT_INDEX (must be a positive integer) "
```
Perhaps keep this line with `(must be a positive integer)` instead of the suggested edit to keep the message consistent across all commands.

wakululuu commented on src/main/java/seedu/address/logic/commands/AssignCommand.java:
```diff
@@ -27,37 +26,33 @@
-            + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer) "
-            + PREFIX_ROLE + "ROLE\n"
+            + PREFIX_SHIFT + "SHIFT_INDEX "
+            + PREFIX_WORKER_ROLE + "WORKER_INDEX ROLE\n"
```
```suggestion
            + "{" + PREFIX_WORKER_ROLE + "WORKER_INDEX (must be a positive integer) ROLE}...\n"
```
`...` would show that the params can be repeated multiple times. I believe there is currently no convention to group a set of command prefixes and params with whitespaces between them together but perhaps we could use `{ }`.

wakululuu commented on src/main/java/seedu/address/logic/commands/AssignCommand.java:
```diff
@@ -27,37 +26,33 @@
 
-    public static final String MESSAGE_ASSIGN_SUCCESS = "New assignment added:\n%1$s";
-    public static final String MESSAGE_DUPLICATE_ASSIGNMENT = "This assignment already exists in the McScheduler";
+    public static final String MESSAGE_ASSIGN_SUCCESS = "%d new assignment(s) added.";
```
Would be good to still see all the assignments added.

wakululuu commented on src/main/java/seedu/address/logic/commands/AssignCommand.java:
```diff
@@ -70,27 +65,31 @@ public CommandResult execute(Model model) throws CommandException {
-        model.updateFilteredWorkerList(PREDICATE_SHOW_ALL_WORKERS);
+            model.addAssignment(assignmentToAdd);
+            model.updateFilteredShiftList(PREDICATE_SHOW_ALL_SHIFTS);
+            model.updateFilteredWorkerList(PREDICATE_SHOW_ALL_WORKERS);
```
I don't think there is a need to update after adding each assignment.

wakululuu commented on src/main/java/seedu/address/logic/commands/UnassignCommand.java:
```diff
@@ -23,32 +24,33 @@
+            + "w/2";
 
-    public static final String MESSAGE_UNASSIGN_SUCCESS = "Shift assignment removed:\n%1$s";
+    public static final String MESSAGE_UNASSIGN_SUCCESS = "%d shift assignment(s) removed.";
```
Similar to `AssignCommand`, it would be good if all assignments removed can be shown.

wakululuu replied:
Similar to `AssignCommand`, it would be good to list all the assignments removed.

wakululuu commented on src/main/java/seedu/address/logic/commands/AssignCommand.java:
```diff
@@ -70,27 +65,31 @@ public CommandResult execute(Model model) throws CommandException {
         }
-        if (workerIndex.getZeroBased() >= lastShownWorkerList.size()) {
-            throw new CommandException(Messages.MESSAGE_INVALID_WORKER_DISPLAYED_INDEX);
+        for (WorkerRole workerRole: workerRoles) {
```
```suggestion
        for (WorkerRole workerRole : workerRoles) {
```

wakululuu commented on src/main/java/seedu/address/logic/commands/AssignCommand.java:
```diff
@@ -70,27 +65,31 @@ public CommandResult execute(Model model) throws CommandException {
-        Worker workerToAssign = lastShownWorkerList.get(workerIndex.getZeroBased());
-        Shift shiftToAssign = lastShownShiftList.get(shiftIndex.getZeroBased());
-        Assignment assignmentToAdd = new Assignment(shiftToAssign, workerToAssign, role);
+        for (WorkerRole workerRole: workerRoles) {
```
```suggestion
        for (WorkerRole workerRole : workerRoles) {
```

wakululuu commented on src/main/java/seedu/address/logic/commands/UnassignCommand.java:
```diff
@@ -61,19 +63,23 @@ public CommandResult execute(Model model) throws CommandException {
         }
-        if (workerIndex.getZeroBased() >= lastShownWorkerList.size()) {
-            throw new CommandException(Messages.MESSAGE_INVALID_WORKER_DISPLAYED_INDEX);
+        for (Index workerIndex: workerIndexes) {
```
```suggestion
        for (Index workerIndex : workerIndexes) {
```

wakululuu commented on src/main/java/seedu/address/logic/commands/UnassignCommand.java:
```diff
@@ -61,19 +63,23 @@ public CommandResult execute(Model model) throws CommandException {
-        Shift shiftToUnassign = lastShownShiftList.get(shiftIndex.getZeroBased());
-        Assignment assignmentToDelete = new Assignment(shiftToUnassign, workerToUnassign);
-        model.deleteAssignment(assignmentToDelete);
+        for (Index workerIndex: workerIndexes) {
```
```suggestion
        for (Index workerIndex : workerIndexes) {
```

wakululuu commented on src/main/java/seedu/address/logic/commands/UnassignCommand.java:
```diff
@@ -61,19 +63,23 @@ public CommandResult execute(Model model) throws CommandException {
 
-        model.updateFilteredShiftList(PREDICATE_SHOW_ALL_SHIFTS);
-        model.updateFilteredWorkerList(PREDICATE_SHOW_ALL_WORKERS);
+            model.updateFilteredShiftList(PREDICATE_SHOW_ALL_SHIFTS);
+            model.updateFilteredWorkerList(PREDICATE_SHOW_ALL_WORKERS);
```
Similar to `AssignCommand`, I don't think we need to update in the loop.

wakululuu commented on src/main/java/seedu/address/model/assignment/WorkerRole.java:
```diff
@@ -0,0 +1,89 @@
+ * Represents a Worker being assigned to a Role for a shift in the App.
+ * Guarantees: details are present and not null, field values are validated, immutable.
+ */
+public class WorkerRole {
```
Personally, `WorkerRole` as a class name doesn't seem very intuitive. Maybe use `WorkerRoleAssignment` or `WorkerRolePair`. The others can chip in on this as well.

wakululuu commented on src/main/java/seedu/address/logic/parser/AssignCommandParser.java:
```diff
@@ -28,25 +28,27 @@
         } catch (IllegalValueException ive) {
             throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT,
-                    AssignCommand.MESSAGE_USAGE), ive);
+                    "AAAA"), ive);
```
Is this a typo?

wakululuu commented on src/main/java/seedu/address/logic/commands/TakeLeaveCommand.java:
```diff
@@ -16,35 +20,40 @@
             + "during the specified shift by the index numbers used in the last worker and shift listings. "
             + "\nParameters: "
             + PREFIX_SHIFT + "SHIFT_INDEX (must be a positive integer) "
             + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)\n"
```
```suggestion
            + "{" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)}...\n"
```

wakululuu commented on src/main/java/seedu/address/logic/commands/UnassignCommand.java:
```diff
@@ -23,32 +24,33 @@
+            + "from the McScheduler by the index numbers used in the last worker and shift listings. "
             + "\nParameters: "
             + PREFIX_SHIFT + "SHIFT_INDEX (must be a positive integer) "
             + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)\n"
```
```suggestion
            + "{" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)}...\n"
```

sigmund-c committed:
misc. text fixes for consistency:
- WorkerRole -> WorkerRolePair
- Assign/Unassign/Leave/Cancel-Leave now lists all the modified assignments
- Cleaner DeveloperGuide
- Consistent naming in command usage text(curly brackets, "...")
- Put model updates outside of loop

sigmund-c committed:
fix output for mass assign/unassign

sigmund-c committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2103-F10-4/tp into mass-Assign

# Conflicts:
#	src/main/java/seedu/address/commons/core/Messages.java
#	src/main/java/seedu/address/logic/commands/AssignCommand.java
#	src/main/java/seedu/address/logic/commands/UnassignCommand.java
#	src/main/java/seedu/address/logic/parser/CliSyntax.java

sigmund-c committed:
fix assign/unassign tests, add WorkerRolePairTest

sigmund-c committed:
fix checkstyle

sigmund-c committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2103-F10-4/tp into mass-Assign

# Conflicts:
#	docs/diagrams/LeaveClassDiagram.puml
#	src/main/java/seedu/address/commons/core/Messages.java
#	src/main/java/seedu/address/logic/commands/AssignCommand.java
#	src/main/java/seedu/address/logic/commands/CancelLeaveCommand.java
#	src/main/java/seedu/address/logic/commands/UnassignCommand.java

sigmund-c committed:
fix tests

sigmund-c committed:
Merge branch 'master' into mass-Assign

sigmund-c committed:
edit unassign response to mention the role too

plosslaw commented on src/main/java/seedu/address/logic/commands/CancelLeaveCommand.java:
```diff
@@ -17,41 +18,43 @@
             + "\nParameters: "
             + PREFIX_SHIFT + "SHIFT_INDEX (must be a positive integer) "
-            + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)\n"
+            + "{" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)}...\n"
```
To be consistent with other commands such as worker-add
```suggestion
            + "[" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)]...\n"
```

plosslaw commented on src/main/java/seedu/address/logic/commands/CancelLeaveCommand.java:
```diff
@@ -64,29 +67,35 @@ public CommandResult execute(Model model) throws CommandException {
+                        workerToCancelLeave, shiftToCancelLeaveFrom));
+            }
+
+            model.deleteAssignment(assignmentInModel);
```
```suggestion
            try {
                model.deleteAssignment(assignmentInModel);
            } catch (AssignmentNotFoundException ex) {
                throw new CommandException(MESSAGE_NO_ASSIGNMENT_FOUND);
            }
```

plosslaw commented on src/main/java/seedu/address/logic/commands/TakeLeaveCommand.java:
```diff
@@ -16,35 +20,40 @@
             + "\nParameters: "
             + PREFIX_SHIFT + "SHIFT_INDEX (must be a positive integer) "
-            + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)\n"
+            + "{" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)}...\n"
```
```suggestion
            + "[" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)]...\n"
```

plosslaw commented on src/main/java/seedu/address/logic/commands/UnassignCommand.java:
```diff
@@ -22,33 +23,36 @@
             + "\nParameters: "
             + PREFIX_SHIFT + "SHIFT_INDEX (must be a positive integer) "
-            + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)\n"
+            + "{" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)}...\n"
```
```suggestion
            + "[" + PREFIX_WORKER + "WORKER_INDEX (must be a positive integer)]...\n"
```

sigmund-c committed:
Update src/main/java/seedu/address/logic/commands/CancelLeaveCommand.java

Co-authored-by: Bohao Guo <37808124+plosslaw@users.noreply.github.com>

sigmund-c committed:
Update src/main/java/seedu/address/logic/commands/UnassignCommand.java

Co-authored-by: Bohao Guo <37808124+plosslaw@users.noreply.github.com>

sigmund-c committed:
Update src/main/java/seedu/address/logic/commands/TakeLeaveCommand.java

Co-authored-by: Bohao Guo <37808124+plosslaw@users.noreply.github.com>

sigmund-c committed:
Update src/main/java/seedu/address/logic/commands/CancelLeaveCommand.java

Co-authored-by: Bohao Guo <37808124+plosslaw@users.noreply.github.com>

sigmund-c committed:
fix missing import

plosslaw reviewed (approved):
LGTM
